### DISCUSSION TITLE
2023-09-13 백엔드 V2 라우트 정리 회의

### DISCUSSION THREAD

**Discussion by scarf005** [jiphyeonjeon-42/backend#767] [General]
참여자 목록 @scarf005 @honeyl3ee @JeongJiHwan @Jiwon-Woo 

## 동사가 없어야 된다 (동사규칙)

HTTP 메서드 (GET, POST, PATCH, DELETE...)만으로 의미를 명확히 표현 가능해야 한다

POST /books/{id} # ok
GET /books/create # ko

## 경로가(`/`) 중요한 순서대로 배치되어야 한다 (중요도규칙)

GET /books/{id}

예시: 내가 작성한 리뷰

중요도 1순위: '내가'
중요도 2순위: '작성한 리뷰'

GET /mypage/reviews/{id} # ok
GET /reviews/mypage/{id} # ko
GET /{id}/reviews/mypage # ko

## 자원은 복수형으로 표기해야 한다 (복수형규칙)

복수와 단수 자료 요청을 명확하게 나타낼 수 있습니다

GET /books/ -> 책 여러권 조회
GET /books/{id} -> 책 한 권 조회

## `id`는 API 경로에 주자 (id규칙)

GET /stock/update -> PUT /stocks/{id}

## 상황에 맞는 메서드를 쓰자 (메서드규칙)

PUT: 기존 자원을 전혀 새로운 자료로 덮어씌움 -> 같은 요청을 여러번 해도 결과가 변하지 않습니다
PATCH: 기존 자원의 값 일부를 고치기 -> 같은 요청을 여러번 하면 결과가 변할 수 있습니다

예시1) 좋아요 취소 -> 좋아요 취소를 여러번 해도 결과가 변하지 않으므로 PATCH보다는 PUT이 적절합니다
예시2) 좋아요 토글 버튼 -> 요청 횟수에 따라 결과가 변하므로 PUT보다는 PATCH가 적절합니다

 
## 오늘의 경로 재작성 목표

![image](https://github.com/jiphyeonjeon-42/backend/assets/54838975/a11cb78d-a98f-4690-9f59-e37ed5e6bfaf)

- [x] stock
- [x] likes
- [x] users

## 로컬에서 swagger 실행하기

![image](https://github.com/jiphyeonjeon-42/backend/assets/54838975/3dbf01ea-e741-4d46-a3fb-271195bd4df0)

1. contracts/src/index.ts 에서 주석처리된 경로 모두 해제
2. 최상위 경로에서 다음 명령어 실행
```sh
pnpm install
pnpm dev
```
3. http://localhost:3000/swagger-v2/

**Comment by scarf005** [comment]
## Stock (제거 후 Books 경로와 통합)

GET `/stock/search` -> GET `/books/`
GET `/stock/update` -> PATCH `/books/{id}`

### `/stock/search` 경로를 /books/로 통합한 이유

1. 도서 검색용 경로인 `/api/v2/books/search` 경로와 목적은 다르나 동작은 동일하기에 (도서 목록 조회), GET /books/ 에 'updatedAt' 기준 조회가 가능하게 해 통일하면 좋을것 같습니다.

### 변경점

1. books로 복수 자료 표현
2. search, update 동사 제거
3. body로 id 단일 인자만 주므로 `/{id}` 경로로 이전

**Comment by Jiwon-Woo** [comment]
### likes
POST /api/v2/books/info/{bookInfoId}/like -> /api/v2/book-infos/{bookInfoId}/like
GET /api/v2/books/info/{bookInfoId}/like -> GET /api/v2/book-infos/{bookInfoId} 통합
DELETE /api/v2/books/info/{bookInfoId}/like -> /api/v2/book-infos/{bookInfoId}/like
- book-info 자체가 리소스인 것 같아서 계층적 구조보다 한 단어로 표현 하면 좋을것 같습니다
- 리소스를 복수형으로 표현하면 좋을 것 같습니다
- GET like 를 GET book-info 에 통합

**Reply by scarf005** [reply]
테스트

**Comment by honeyl3ee** [comment]
users

GET `/api/v2/users/search` -> GET `/api/v2/users`
POST `/v2/users/create` -> POST `/v2/users`
PATCH `/v2/users/update/{id}` -> PATCH `/v2/users/{id}`

* 동사 제거

**Comment by JeongJiHwan** [comment]
GET `books/create` -> `books/generation`
- 책 생성을 위한 정보를 반환하는 route라서 기존의 create를 명사형으로 변환하였습니다.

GET `books/search` -> `books`
POST `books/create` -> `books`
PATCH `books/update` -> `books` & `book-infos` 현재 books 데이터와 book-info 데이터가 하나의 route에서 동작하도록 되어있어 books와 book-info를 분리하면 좋을 것 같습니다.
PATCH `books/donator` -> `books` 와 body 데이터만 추가하여 합치는 방향이 좋을 것 같습니다.

GET `books/info/{id}` -> `book-infos/{id}`
GET `books/info/search` -> `book-infos`
GET `books/info/tag` -> `book-infos/tag`
GET `books/info` -> `book-infos/sorted`

**Comment by Jiwon-Woo** [comment]
### histories
GET /mypage/histories -> /mypage/lendings
GET /histories -> /lendings
- 정보를 lending 테이블에서 가져오고 대출에 대한 기록이므로 histories 대신 lendings로 하는 건 어떨까요?

**Comment by honeyl3ee** [comment]
users에서 
UPDATE `/v2/users/myupdate` 가 어떤 역할인지 모르겠습니다.

**Reply by scarf005** [reply]
아이디, 비밀번호 업데이트용 경로인데, 저희가 소셜 로그인으로 전환중이기 때문에 구현에서 제외해도 좋을 것 같습니다.

**Comment by Jiwon-Woo** [comment]
GET /api/v2/books/info/{bookInfoId}/like
PATCH /v2/users/myupdate
GET books/info/tag -> book-infos/tag
GET books/info -> book-infos/sorted

위 라우트는 삭제해도 괜찮겠다고 얘기 나온 것들입니다.

**Comment by scarf005** [comment]
## book과 bookinfo 명칭 고도화

book_info -> bookmeta
